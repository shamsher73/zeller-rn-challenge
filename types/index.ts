export type User = {
    id: string,
    name: string,
    role: string,
}

export enum UserType {
    ADMIN = 'ADMIN',
    MANAGER = 'MANAGER',
}

export type QueryResponse = {
    data: {
        listZellerCustomers: {
            items: User[];
        };
    }
};