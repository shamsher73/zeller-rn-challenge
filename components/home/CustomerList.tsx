import { View, FlatList, StyleSheet, Text, RefreshControl } from 'react-native';
import ListItem from './ListItem';
import { ThemedText } from '../ThemedText';
import { User, UserType } from '@/types';
import { capitalizeWords } from '@/utils';
import Divider from '../Divider';
import { useCallback, useState } from 'react';

type CustomerListProps = {
    items: User[],
    selectedValue: UserType
    getUsers: () => void
}

export default function CustomerList(props: CustomerListProps) {
    const { items, selectedValue, getUsers } = props
    const [refreshing, setRefreshing] = useState(false);

    const onRefresh = useCallback(() => {
        setRefreshing(true);
        getUsers()
        setTimeout(() => {
            setRefreshing(false);
        }, 2000);
    }, []);

    return (
        <View>
            <ThemedText type='title'>{capitalizeWords(selectedValue)} Users</ThemedText>
            <FlatList
                style={styles.listContainer}
                data={items}
                renderItem={(item) => (selectedValue === item.item.role ? <ListItem user={item.item} /> : null)}
                ListEmptyComponent={() => <Text>No users found</Text>}
                refreshControl={
                    <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
                }
            />
            <Divider />
        </View>
    );
}

const styles = StyleSheet.create({
    listContainer: {
        marginVertical: 20,

    }
});
