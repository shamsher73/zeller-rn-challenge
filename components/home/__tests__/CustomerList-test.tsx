import { render } from '@testing-library/react-native';
import CustomerList from '../CustomerList';
import { User, UserType } from '@/types';

describe('<CustomerList />', () => {

    it('should render a list of users based on the selected role', () => {
        const items = [
            { id: '1', name: 'John Doe', role: 'ADMIN' },
            { id: '2', name: 'Jane Smith', role: 'ADMIN' }
        ];
        const selectedValue = UserType.ADMIN;
        const getUsers = jest.fn()
        const { getByText } = render(<CustomerList items={items} selectedValue={selectedValue} getUsers={getUsers} />);
        expect(getByText('Admin Users')).toBeTruthy();
        expect(getByText('John Doe')).toBeTruthy();
        expect(getByText('Jane Smith')).toBeTruthy();
    });

    it('should display role in customer list based on selected role', () => {
        const items = [
            { id: '1', name: 'John Doe', role: 'ADMIN' },
            { id: '2', name: 'Jane Smith', role: 'MANAGER' }
        ];
        const selectedValue = UserType.ADMIN;
        const getUsers = jest.fn()
        const { getByText } = render(<CustomerList items={items} selectedValue={selectedValue} getUsers={getUsers} />);
        expect(getByText('Admin Users')).toBeTruthy();
        expect(getByText('Admin')).toBeTruthy();
    });


    it('should handle empty list of users gracefully', () => {
        const items: User[] = [];
        const selectedValue = UserType.MANAGER;
        const getUsers = jest.fn()
        const { getByText, queryByText } = render(<CustomerList items={items} selectedValue={selectedValue} getUsers={getUsers} />);
        expect(getByText('Manager Users')).toBeTruthy();
        expect(queryByText('No users found')).toBeTruthy();
    });

    test('snapshot', () => {
        const items = [
            { id: '1', name: 'John Doe', role: 'ADMIN' },
            { id: '2', name: 'Jane Smith', role: 'ADMIN' }
        ];
        const selectedValue = UserType.ADMIN;
        const getUsers = jest.fn()
        const tree = render(<CustomerList items={items} selectedValue={selectedValue} getUsers={getUsers} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
