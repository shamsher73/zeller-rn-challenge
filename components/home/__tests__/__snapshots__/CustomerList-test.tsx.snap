// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomerList /> snapshot 1`] = `
<View>
  <Text
    style={
      [
        undefined,
        {
          "fontSize": 22,
          "fontWeight": "500",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Admin
     Users
  </Text>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={
      [
        {
          "id": "1",
          "name": "John Doe",
          "role": "ADMIN",
        },
        {
          "id": "2",
          "name": "Jane Smith",
          "role": "ADMIN",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "marginVertical": 20,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
              "marginVertical": 10,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#e8f2fb",
                "borderRadius": 8,
                "height": 40,
                "justifyContent": "center",
                "width": 40,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#0370ce",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": 800,
                }
              }
            >
              J
            </Text>
          </View>
          <View
            style={
              {
                "justifyContent": "space-between",
                "marginLeft": 10,
              }
            }
          >
            <Text
              style={
                [
                  undefined,
                  undefined,
                  {
                    "fontSize": 16,
                    "fontWeight": "400",
                  },
                  undefined,
                  undefined,
                ]
              }
            >
              John Doe
            </Text>
            <Text
              style={
                [
                  undefined,
                  undefined,
                  undefined,
                  {
                    "fontSize": 12,
                    "fontWeight": "300",
                  },
                  undefined,
                ]
              }
            >
              Admin
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
              "marginVertical": 10,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#e8f2fb",
                "borderRadius": 8,
                "height": 40,
                "justifyContent": "center",
                "width": 40,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#0370ce",
                  "fontSize": 18,
                  "fontStyle": "normal",
                  "fontWeight": 800,
                }
              }
            >
              J
            </Text>
          </View>
          <View
            style={
              {
                "justifyContent": "space-between",
                "marginLeft": 10,
              }
            }
          >
            <Text
              style={
                [
                  undefined,
                  undefined,
                  {
                    "fontSize": 16,
                    "fontWeight": "400",
                  },
                  undefined,
                  undefined,
                ]
              }
            >
              Jane Smith
            </Text>
            <Text
              style={
                [
                  undefined,
                  undefined,
                  undefined,
                  {
                    "fontSize": 12,
                    "fontWeight": "300",
                  },
                  undefined,
                ]
              }
            >
              Admin
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "borderBottomColor": "grey",
        "borderBottomWidth": 0.5,
        "marginVertical": 16,
      }
    }
  />
</View>
`;
